Rewrite for RN 0.50+

jasonfma opened the pull request:
Major upgrade that supports 0.50+
- support by adding `reactNativeVersion` and `DeviceInfo` properties to the `NativeModules`
- Remove `Navigator` test as it was removed from RN
- Fix `BackAndroid` and add `BackHandler` tests
- update all packages
- fix lint issues
- drop support for Node 4

NOTE: This is a PR into the existing `rewrite` branch

Commit:
Update to RN 0.49
- update enzyme and use enzyme's react 16 adapter
- add RN and react devDeps for now (reflects the peerDeps)
- Add newly required `reactNativeVersion` and `DeviceInfo` properties to the `NativeModules`
  - `reactNativeVersion` reads from the RN source
  - `DeviceInfo` is set to iPhone 6

Commit:
Fix BackAndroid and add BackHandler
- BackAndroid is deprecated by BackHandler
- added test for BackHandler
- Remove obsolete test of `addEventListener().remove()`

Commit:
Remove `Navigator` test as it was removed from RN

Commit:
Fix Dimensions test

Commit:
Fix tests for Android progress bar and NativeModules Dimensions

Commit:
Update for RN 0.50
- backwards compatible changes for RN 0.50
- RN 0.50 adds attributes to components and breaks the buildComponentHTML approach to testing

Commit:
Package updates
- TODO lint fixes

Commit:
Automatic lint fixes

Commit:
Manual lint fixes

Commit:
Fix test broken by chai upgrade
- remove unnecessary assertion because it's asserted implicitly in the next assertion

Commit:
More package updates

Commit:
Update travis to test RN 0.49+

Signed-off-by: Jason Ma <jason@peartherapeutics.com>

Commit:
Remove react-addons-test-utils from travis builds

Commit:
Use RN 0.50+
- render of `ProgressBarAndroid` changed from `ActivityIndicator` in RN 0.49 back to  `AndroidProgressBar` in RN 0.50

Signed-off-by: Jason Ma <jason@peartherapeutics.com>

jasonfma commented:
Changing to support RN 0.50+ because the `AndroidProgressBar` had a breaking change in RN 0.49 that was fixed in RN 0.50. Rather than start of backwards compatible with 0.49, just do 0.50+.

Commit:
Add tests for Node 8 now that it's LTS

Signed-off-by: Jason Ma <jason@peartherapeutics.com>

Commit:
Add back devDeps for RN and react to fix build

Signed-off-by: Jason Ma <jason@peartherapeutics.com>

Commit:
Fix CI to not rely on devDeps react and react-native

Signed-off-by: Jason Ma <jason@peartherapeutics.com>

Commit:
Drop support for Node 4

Signed-off-by: Jason Ma <jason@peartherapeutics.com>

jasonfma commented:
Dropping support for Node 4. https://travis-ci.org/RealOrangeOne/react-native-mock/jobs/297785954

dustinsavery commented:
Should probably add `react`, `react-dom`, and `react-native` as dev dependencies also. When trying to build locally without those packages, it fails

RealOrangeOne commented:
Initial skim looks good. Would probably be ideal to update the `README` stating that we've dropped support for Node `<6` (and why?). Will give it a proper review soon, but glad to see the tests are passing!

gabceb commented:
Looking forward to this @RealOrangeOne and @jasonfma. When I try to pull this branch as a dev dependency for my project running RN 0.50 and React 16 it complains when running haste.js as part of the post install

```
Error: EISDIR: illegal operation on a directory, read
    at Error (native)
    at Object.fs.readSync (fs.js:731:19)
    at tryReadSync (fs.js:486:20)
    at Object.fs.readFileSync (fs.js:534:19)
    at /Users/gabrielcebrian/Projects/tes/mobile/app-job-mobile/node_modules/react-native-mock/src/haste.js:32:39
    at Function._.each._.forEach (/Users/gabrielcebrian/Projects/tes/mobile/app-job-mobile/node_modules/underscore/underscore.js:153:9)
    at Object.<anonymous> (/Users/gabrielcebrian/Projects/tes/mobile/app-job-mobile/node_modules/react-native-mock/src/haste.js:31:3)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
```

If I clone the project on a different folder the post-install finishes as expected

EdenGottlieb commented:
Adding a mock for FlatList would be a great idea, in my opinion.

RealOrangeOne commented on .travis.yml:
```diff
@@ -1,22 +1,16 @@
   - "6.0.0"
-  - "4"
   - "6"
+  - "8"
```
Could we keep the base versions here too, so add `8.0.0`

dustinsavery replied:
Sure, I can add that

jasonfma replied:
Should we use `8.9.1` as that is the LTS version?

RealOrangeOne replied:
Using `8` specifically targets the latest release at build time, which is fine. I quite like the idea of testing against the base version as well

RealOrangeOne commented on tests/integration/apis/Easing.test.js:
```diff
@@ -14,7 +14,7 @@ import { expect } from 'chai';
   const { Easing } = require('react-native');
 
-  it('should have all functions', function () {
+  it('should have all functions', () => {
```
Please revert changes like this back. `function` is correct here

dusave replied:
Arrow functions should be just fine. You're using them above on line 14 already. Is there something I'm missing?

jasonfma replied:
Sorry, this change was part of the eslint rule update. I just wholesale applied the rule fixes. I can revert, but do you have more context as to why this one needs to be function?

RealOrangeOne replied:
I assumed all were using `function`, in that case this is totally fine

RealOrangeOne reviewed (changes requested):
Overall looks pretty good, a couple small comments. Amazing work! 🎉

jasonfma commented:
@RealOrangeOne just to be clear, this is a PR into the `rewrite` branch that you created. This is not a merge into master.

Commit:
Node support
- Update readme to reference Node 6 support
- Update travis to include Node 8.0.0 tests

jasonfma commented:
@RealOrangeOne Looks like node 8.0.0 is broken. I'm going to set to 8.9.1 since it's the first LTS version.

RealOrangeOne commented:
Just seen the build log, agreed that's very weird. I think sticking to that version will probably be fine, it's early enough!

Commit:
Set base Node 8 test version to 8.9.1

jasonfma commented:
@RealOrangeOne can you rebuild? I think it failed because of https://github.com/npm/registry/issues/255

jasonfma commented:
> dustinsavery commented on Nov 9, 2017
> Should probably add react, react-dom, and react-native as dev dependencies also. When trying to build locally without those packages, it fails

I think this is set up this way so the build can test with various versions of each. I did add them as `peerDeps` so you can see the warnings.

@RealOrangeOne comments addressed

jasonfma commented:
@RealOrangeOne will you have some time soon to take a look? I'm happy to help here where I can too. This PR is to merge into your existing `rewrite` branch #120 where there are still some merge conflicts to resolve as well - so this is not final and going into `master`.

RealOrangeOne commented:
All these changes look good to me! Let's get this into `rewrite` and see how it affects it!

lalop commented:
I tried rewrite branch but I get this error 
```
Cannot find module './build/react-native-mock.js' from 'mock.js'
```

Will we use this branch as master ?

RealOrangeOne commented:
If you're downloading straight from GitHub, you'll have to build the compiled code to actually use it. 

Yes, this code will eventually be pushed into master. Time frame unknown, but shouldn't be too long!

lalop commented:
I get it via yarn, but build folder isn't present.
Is that normal ?

RealOrangeOne commented:
Yarn, npm, makes no difference. We don't commit the `build/` directory for obviousl reasons, committing build artifacts is bad practice. If you want to use a version from somewhere other than npm, you'll need to run the build yourself from the project directory, else you'll only have the source files, which won't work.

lalop commented:
I'm agree with you but I'm not sure to understand.
I get react-native-mock via yarn, that provide source code not built but https://github.com/RealOrangeOne/react-native-mock/blob/rewrite/mock.js call `./build/react-native-mock.js` so is there something wrong here or should I build the package even if I get it via npm/yarn ?

RealOrangeOne commented:
hmm, this sounds odd, and probably not the best place to discuss. Perhaps raise an actual issue so it can be looked into

lalop commented:
done https://github.com/RealOrangeOne/react-native-mock/issues/157
